lib/at_parser/at_parser.c

```diff
@@ -0,0 +1,411 @@
+		return -EPERM;
+	}
+
+	ret = at_parser_seek(parser, index, &token);
+	if (ret) {
+		return ret;
+	}
+
+	if (token.type != AT_TOKEN_TYPE_INT) {
+		return -EINVAL;
+	}
+
+	errno = 0;
+	char *next = NULL;
+	int64_t val = strtoll(token.start, &next, 10);
+
+	if (errno == ERANGE) {
+		return -ERANGE;
+	}
+
+	switch (type) {
+	case AT_NUM_INT16:
+		if (val < INT16_MIN || val > INT16_MAX) return -ERANGE;
+		*(int16_t *)value = (int16_t)val;
+		break;
+	case AT_NUM_UINT16:
+		if (val < 0 || val > UINT16_MAX) return -ERANGE;
+		*(uint16_t *)value = (uint16_t)val;
+		break;
+	case AT_NUM_INT32:
+		if (val < INT32_MIN || val > INT32_MAX) return -ERANGE;
+		*(int32_t *)value = (int32_t)val;
+		break;
+	case AT_NUM_UINT32:
+		if (val < 0 || val > UINT32_MAX) return -ERANGE;
+		*(uint32_t *)value = (uint32_t)val;
+		break;
+	case AT_NUM_INT64:
+		*(int64_t *)value = val;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int at_parser_int16_get(struct at_parser *parser, size_t index, int16_t *value) {
+	return at_parser_num_get_impl(parser, index, value, AT_NUM_INT16);
+}
+
+int at_parser_uint16_get(struct at_parser *parser, size_t index, uint16_t *value) {
+	return at_parser_num_get_impl(parser, index, value, AT_NUM_UINT16);
+}
+
+int at_parser_int32_get(struct at_parser *parser, size_t index, int32_t *value) {
+	return at_parser_num_get_impl(parser, index, value, AT_NUM_INT32);
+}
+
+int at_parser_uint32_get(struct at_parser *parser, size_t index, uint32_t *value) {
+	return at_parser_num_get_impl(parser, index, value, AT_NUM_UINT32);
+}
+
+int at_parser_int64_get(struct at_parser *parser, size_t index, int64_t *value) {
+	return at_parser_num_get_impl(parser, index, value, AT_NUM_INT64);
+}
+
+int at_parser_uint64_get(struct at_parser *parser, size_t index, uint64_t *value) {
+	/* todo: update with UINT64 */
+	return at_parser_num_get_impl(parser, index, value, AT_NUM_INT64);
+}
+
+int at_parser_string_get(struct at_parser *parser, size_t index, char *str, size_t *len)
+{
+	int ret;
+	struct at_token token = {0};
+
+	if (!parser || !str || !len || !parser->cursor) {
+		return -EINVAL;
+	}
+
+	if (!parser->is_init) {
+		return -EPERM;
+	}
+
+	ret = at_parser_seek(parser, index, &token);
+	if (ret) {
+		return ret;
+	}
+
+	switch (token.type) {
+	/* Acceptable types. */
+	case AT_TOKEN_TYPE_CMD_TEST:
+	case AT_TOKEN_TYPE_CMD_SET:
+	case AT_TOKEN_TYPE_CMD_READ:
+	case AT_TOKEN_TYPE_NOTIF:
+	case AT_TOKEN_TYPE_QUOTED_STRING:
+	case AT_TOKEN_TYPE_STRING:
+	case AT_TOKEN_TYPE_ARRAY:
+	case AT_TOKEN_TYPE_RESP:
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* Check if there is enough memory for null terminator. */
+	if (*len < token.len + 1) {
+		return -ENOMEM;
+	}
+
+	memcpy(str, token.start, token.len);
+
+	/* Null-terminate the output string. */
+	str[token.len] = '\0';
+
+	/* Update the length to reflect the copied string length. */
+	*len = token.len;
+
+	return 0;
+}
+
+int at_parser_next_line_move(struct at_parser *parser)
+{
+	int ret;
```

Comment:
nit: I prefer `err` to `ret`, especially when the return value is only an error and not an integer counting something.

Then the loop below would be a bit clearer:

```
do {
	ret = at_parser_tok(parser, &token);
} while (ret == 0);

```
becomes
```
do {
	err = at_parser_tok(parser, &token);
} while (!err);
```